icicle-compiler/src/Icicle/Sea/Eval/Base.hs

```diff
@@ -255,64 +257,67 @@ seaCompile' options cache input programs chords = do
+          return $ \f ptr -> withCPiano fp (f ptr)
 
   take_snapshot <- case input of
     NoInput -> do
-      return (\_ -> return ())
+      return $ \_ -> return ()
 
-    HasInput (FormatPsv _) _ -> do
+    HasInput (FormatPsv _) _ _ -> do
       fn <- firstEitherT SeaJetskiError (function lib "psv_snapshot" retVoid)
       return $ play $ \ptr cpiano -> fn [ argPtr (unCPiano cpiano), argPtr ptr ]
 
-    HasInput (FormatZebra _ _) _ -> do
+    HasInput (FormatZebra _ _) _ input_path -> do
       step <- firstEitherT SeaJetskiError (function lib "zebra_snapshot_step" (retPtr retCChar))
       init <- firstEitherT SeaJetskiError (function lib "zebra_alloc_state" (retPtr retVoid))
 
-      -- FIXME this is really savage.
-      let go ptr cpiano = do
-            let piano   = unCPiano cpiano
-            input_ptr  <- peekWordOff ptr 0
-            input_path <- peekCString input_ptr
+      (puller, pullid) <- hoist liftIO
+                        $ firstEitherT (SeaExternalError . T.pack . show)
+                        $ blockChainPuller (VB.singleton input_path)
 
-            pull <- runEitherT (blockChainPuller (VB.singleton input_path))
+      let withPiano :: Ptr a -> CPiano -> IO ()
+          withPiano ptr cpiano = do
+            let piano = unCPiano cpiano
 
-            case pull of
-              Left e ->
-                Savage.error $ "error creating puller: " <> show e
+            state <- init [ argPtr piano, argPtr ptr ]
 
-              Right (puller, pullid) -> do
-                st <- init [ argPtr piano, argPtr ptr ]
+            let step' :: CEntity -> EitherT SeaError IO ()
+                step' e = do
+                  s <- liftIO $ step [ argPtr piano, argPtr state, argPtr (unCEntity e) ]
+                  if s /= nullPtr
+                  then do
+                    msg <- liftIO $ peekCString s
+                    left . SeaExternalError . T.pack $ "error step: " <> show msg
+                  else return ()
 
-                let step' e = liftIO $ do
-                      s <- step [ argPtr piano, argPtr st, argPtr (unCEntity e) ]
-                      when (s /= nullPtr) $ do
-                        msg <- liftIO (peekCString s)
-                        Savage.error $ "error step: " <> msg
+            let puller' :: PullId -> EitherT SeaError IO (Maybe Block)
+                puller' = firstEitherT (SeaExternalError . T.pack . show) . puller
 
-                ret <- runEitherT
-                     $ joinErrors show show
-                     $ mergeBlocks (MergeOptions puller step' 100) pullid
+            ret <- runEitherT . joinErrors show show
+                 $ mergeBlocks (MergeOptions puller' step' 100) pullid
 
-                case ret of
-                  Left e ->
-                    Savage.error $ "error merge: " <> show e
-                  Right _ ->
-                    return ()
+            -- need to error straight from IO as piano is tied to IO
```

Comment:
`withPiano` is the continuation that will be given to `withCPiano :: ForeignPiano -> (CPiano -> IO a) -> IO a`